File: DSA/src/Questions/Array/MergeSort.java
package Questions.Array;

public class MergeSort {
    public static void merge(int[] arr, int l, int m, int r) {
        // Find the size of two subarrays to be merged
        int n1 = m - l + 1, n2 = r - m;
        // Create two temporary arrays
        int[] left = new int[n1], right = new int[n2];
        for (int i = 0; i < n1; i++) {
            left[i] = arr[l + i];
        }
        for (int i = 0; i < n2; i++) {
            right[i] = arr[m + 1 + i];
        }
        // Merge the two sorted arrays
        int i = 0, j = 0, k = l;
        while (i < n1 && j < n2) {
            if (left[i] <= right[j]) {
                arr[k++] = left[i++];
            } else {
                arr[k++] = right[j++];
            }
        }
        while (i < n1) {
            arr[k++] = left[i++];
        }
        while (j < n2) {
            arr[k++] = right[j++];
        }
    }

    public static void mergeSort(int[] arr, int l, int r) {
        if (l < r) {
            int m = l + (r - l) / 2;
            mergeSort(arr, l, m);
            mergeSort(arr, m + 1, r);
            merge(arr, l, m, r);
        }
    }

    public static void main(String[] args) {
        int[] arr = { 12, 11, 13, 5, 6, 7 };
        mergeSort(arr, 0, arr.length - 1);
        for (int i : arr) {
            System.out.print(i + " ");
        }
    }
}

File: DSA/Python/Questions/Array/MergeSort.py
"""
# Question: Merge Sort Implementation
# Classic divide and conquer sorting algorithm

# Time Complexity: O(n log n)
# Space Complexity: O(n) for temporary arrays

# Algorithm:
# 1. Divide array into two halves recursively
# 2. Sort each half independently
# 3. Merge sorted halves into single sorted array

# Key Components:
# - merge(): Combines two sorted subarrays
# - merge_sort(): Main recursive function
# - In-place merging with temporary arrays
"""

from typing import List


class MergeSort:
    def merge(self, arr: List[int], l: int, m: int, r: int) -> None:
        # Find sizes of two subarrays
        n1, n2 = m - l + 1, r - m

        # Create temporary arrays
        left = [arr[l + i] for i in range(n1)]
        right = [arr[m + 1 + i] for i in range(n2)]

        # Merge temporary arrays back into arr[l..r]
        i = j = 0
        k = l

        while i < n1 and j < n2:
            if left[i] <= right[j]:
                arr[k] = left[i]
                i += 1
            else:
                arr[k] = right[j]
                j += 1
            k += 1

        # Copy remaining elements if any
        while i < n1:
            arr[k] = left[i]
            i += 1
            k += 1

        while j < n2:
            arr[k] = right[j]
            j += 1
            k += 1

    def merge_sort(self, arr: List[int], l: int, r: int) -> None:
        if l < r:
            m = (l + r) >> 1
            self.merge_sort(arr, l, m)
            self.merge_sort(arr, m + 1, r)
            self.merge(arr, l, m, r)


def main():
    solution = MergeSort()
    arr = [12, 11, 13, 5, 6, 7]
    solution.merge_sort(arr, 0, len(arr) - 1)
    print(arr)  # Expected output: [5, 6, 7, 11, 12, 13]


if __name__ == "__main__":
    main()
